packages/json-schema/test/integration/build-schema.test.ts

```diff
@@ -0,0 +1,159 @@
+import {
+  modelToJSONDef,
+  JSONDefinition,
+  toJSONProperty,
+} from '../../src/build-schema';
+import {expect} from '@loopback/testlab';
+
+describe('build-schema', () => {
+  context('exception', () => {
+    it('errors out when "@property.array" is not used on an array', () => {
+      @model()
+      class BadArray {
+        @property() badArr: string[];
+      }
+      expect(() => {
+        toJSONProperty(modelToJSONDef(BadArray));
+      }).to.throw(/type is defined as an array/);
+    });
+  });
+  describe('modelToJSON', () => {
+    @model()
+    class Bar {
+      @property() barA: number;
+    }
+    @model()
+    class Foo {
+      @property() str: string;
+      @property() num: number;
+      @property() bool: boolean;
+      @property() obj: object;
+      @property() nul: null;
+      @property() undef: undefined;
+      @property.array(String) arrStr: string[];
+      @property() bar: Bar;
+      @property.array(Bar) arrBar: Bar[];
+    }
+    let fooJSON: JSONDefinition;
+    let expected: JSONDefinition;
+    before(() => {
+      fooJSON = modelToJSONDef(Foo);
+    });
+    it('does not convert null or undefined property', () => {
+      expect(fooJSON.properties).to.not.containEql({
+        nul: {type: 'null'},
+      });
+      expect(fooJSON.properties).to.not.containEql({
+        undef: {type: 'undefined'},
+      });
+    });
+    it('does not convert properties that have not been decorated', () => {
+      @model()
+      class NoPropertyMeta {
+        prop: string;
+      }
+      @model()
+      class OnePropertyDecorated {
+        @property() foo: string;
+        bar: boolean;
+        baz: number;
+      }
+      expect(modelToJSONDef(NoPropertyMeta)).to.eql({});
+      expect(modelToJSONDef(OnePropertyDecorated)).to.deepEqual({
+        properties: {
+          foo: {
+            type: 'string',
+          },
+        },
+      });
+    });
+    it('does not convert models that have not been decorated with @model()', () => {
+      class Empty {}
+      class NoModelMeta {
+        @property() foo: string;
+        bar: number;
+      }
+      expect(modelToJSONDef(Empty)).to.eql({});
+      expect(modelToJSONDef(NoModelMeta)).to.eql({});
+    });
+    it('properly converts to JSON definition', () => {
+      expect(fooJSON).to.deepEqual({
+        properties: {
+          str: {
+            type: 'string',
+          },
+          num: {
+            type: 'number',
+          },
+          bool: {
+            type: 'boolean',
+          },
+          obj: {
+            type: 'object',
+          },
+          arrStr: {
+            type: 'array',
+            items: {
+              type: 'string',
+            },
+          },
+          bar: {
+            $ref: '#definitions/Bar',
+          },
+          arrBar: {
+            type: 'array',
+            items: {
+              $ref: '#definitions/Bar',
+            },
+          },
+        },
```

Comment:
I am concerned that when this test fails, the error message will be difficult to understand, because the expected and actual objects are so large!

Please split this test into several smaller tests, where each test is checking one of the types.

```ts
it('properly converts "string" properties', () => {
  @model()
  class TestModel {
    @property() str: string;
  }

  const jsonDef = modelToJSONDef(TestModel);
  expect(jsonDef.properties).to.deepEqual({
    str: {
      type: 'string',
    },
  });
});
```

It is also important to make these tests resilient to future changes. For example, when `modelToJSONDef` starts emitting more top-level fields beyond `properties`, then we don't want to have to update all tests in this file with the same default values for these new properties. My example code above is already addressing that problem by verifying only the contents of `jsonDef.properties`. There should be a dedicated test to verify the overall shape of `jsonDef`.

See also "Isolation of failures" and "Resilience" in http://loopback.io/doc/en/lb4/Defining-your-testing-strategy.html:

> - **Isolation of failures:** The test suite should make it easy to isolate the source of test failures. To fix a failing test, developers need to find the specific place that does not work as expected. When the project contains thousands of lines and the test failure can be caused by any part of the system, then finding the bug is very difficult, time consuming and demotivating.
> - **Resilience:** The test implementation should be robust and resilient to changes in the tested code. As the project grows and matures, you may need to change existing behavior. With a brittle test suite, each change may break dozens of tests, for example when you have many end-to-end/UI tests that rely on specific UI layout. This makes change prohibitively expensive, up to a point where you may start questioning the value of such test suite.
References: